ocaml/xapi/rbac.ml

```diff
@@ -114,86 +110,55 @@ let permission_of_action ?args ~keys _action =
-            (List.fold_left (fun ss s -> ss ^ s ^ ",") "" arg_keys)
-            (List.fold_left
-               (fun ss s -> ss ^ Rpc.to_string s ^ ",")
-               "" arg_values
-            ) ;
-          get_keyERR_permission_name action "DENY_WRGLEN" (* will always deny *)
-        ) else (* keys and values have the same length *)
-          let rec get_permission_name_of_keys arg_keys arg_values =
-            match (arg_keys, arg_values) with
-            | [], [] | _, [] | [], _ ->
-                (* this should never happen *)
-                debug "DENYING access: no 'key' argument in the action %s"
-                  action ;
-                get_keyERR_permission_name action "DENY_NOKEY"
-                (* deny by default *)
-            | k :: ks, v :: vs -> (
-                if k <> "key" (* "key" is defined in datamodel_utils.ml *) then
-                  get_permission_name_of_keys ks vs
-                else (* found "key" in args *)
-                  match v with
-                  | Rpc.String key_name_in_args -> (
-                    try
-                      (*debug "key_name_in_args=%s, keys=[%s]" key_name_in_args ((List.fold_left (fun ss s->ss^s^",") "" keys)) ;*)
-                      let key_name =
-                        List.find
-                          (fun key_name ->
-                            if Astring.String.is_suffix ~affix:"*" key_name then
-                              (* resolve wildcards at the end *)
-                              Astring.String.is_prefix
-                                ~affix:
-                                  (String.sub key_name 0
-                                     (String.length key_name - 1)
-                                  )
-                                key_name_in_args
-                            else (* no wildcards to resolve *)
-                              key_name = key_name_in_args
-                          )
-                          keys
-                      in
-                      get_key_permission_name action
-                        (String.lowercase_ascii key_name)
-                    with Not_found ->
-                      (* expected, key_in_args is not rbac-protected *)
-                      action
-                  )
-                  | _ ->
-                      (* this should never happen *)
-                      debug
-                        "DENYING access: wrong XML value [%s] in the 'key' \
-                         argument of action %s"
-                        (Rpc.to_string v) action ;
-                      get_keyERR_permission_name action "DENY_NOVALUE"
+          debug "DENYING access: no 'key' argument in the action %s" action ;
+          get_keyERR_permission_name action "DENY_NOKEY"
+      | Some (Rpc.String key_name_in_args) -> (
+        try
+          let key_name =
+            List.find
+              (fun key_name ->
+                if Astring.String.is_suffix ~affix:"*" key_name then
```

Comment:
Would it be faster to do this with a regular expression because it could avoid repeated traversal?

Reply:
There's a traversal on key_name (to look for '*'), and another in key_name_in_args. Not sure about the "repeated traversal"